doc/dev/code-generator.md

```diff
@@ -0,0 +1,211 @@
+all templates. Let's dissect this line:
+
+- `ExampleGenTemplate ()` we declare a zero argument constructor. All templates _must_ have only a single zero argument constructor,
+  otherwise the framework cannot instantiate them.
+- `: GenTemplate (` the constructor must invoke the base-class constructor.
+- `"example",` we decided to use `example` as the base name for our template files. This base name will be replaced by the `outputName`
+  chosen by the user, when invoking the code-generator.
+- `{ ".txt" },` this is the list of part suffixes for our template. We only have a single part (output file) with the suffix `.txt`.
+- `{},` the first empty list would contain all the partials our template uses. We don't use any.
+- `{},` the second empty list, is actually a map. It would contain all parameters and whether they are required or not.
+  We don't have parameters.
+
+As you can see, the header is quite simple. The more important part is actually the source file. It contains the implementation of
+`getTemplateData`, the function that delivers all the template data to the framework.
+
+For our example we will use this implementation:
+
+```cpp
+kainjow::mustache::data ExampleGenTemplate::getTemplateData (const std::string & outputName, const std::string & part,
+                                                             const kdb::KeySet & ks, const std::string & parentKey) const
+{
+    using namespace kdb;
+    using namespace kainjow::mustache;
+
+    list keyList;
+
+    for (auto it = ks.begin (); it != ks.end (); ++it)
+    {
+        Key key = *it;
+
+        auto keyObject = object({ { "name", key.getName() } });
+
+        keyList.emplace_back(keyObject);
+    }
+
+    auto data = object({ { "keys", keyList } });
+
+    return data;
+}
+```
+
+The framework will invoke `getTemplateData` for each part of our template with the `outputName` and `parentKey` as given on the command line,
+as well as the current part suffix (`part`) and the input keyset `ks`. All keys of the input keyset are guaranteed to be below `parentKey`.
+
+How exactly the above code works is left as an exercise to the reader.
+
+## Adding the class to `GenTemplateList`
+
+To make the framework aware of our class (and by extension our template), we have to then add it to `GenTemplateList`. This is simply done
+by adding a line to the implementation of `GenTemplateList::GenTemplateList ()` in [`template.cpp`](/src/tools/kdb/gen/template.cpp).
+
+For our example that is:
+
+```cpp
+addTemplate<ExampleGenTemplate> ("example");
+```
+
+We need to specify `"example"` again, because this string defines how our template shall be called, i.e. what we need to specify in the
+terminal to invoke it.
+
+## Using the new template
+
+Now you should be able to use the new template by running (after compiling/installing Elektra again):
+
+```
+kdb gen example user userkeys
```

Comment:
In which folder are we now?

Reply:
Doesn't matter. The command should be equivalent to `kdb ls user > userkeys.txt`, which also doesn't care about cwd.